File: app/src/main/java/com/example/pam_app/di/ProductionContainer.java
package com.example.pam_app.di;

import android.content.Context;

import com.example.pam_app.db.BucketDao;
import com.example.pam_app.db.IncomeDao;
import com.example.pam_app.db.WallyDatabase;
import com.example.pam_app.repository.BucketMapper;
import com.example.pam_app.repository.BucketRepository;
import com.example.pam_app.repository.IncomeMapper;
import com.example.pam_app.repository.IncomeRepository;
import com.example.pam_app.repository.LanguagesRepository;
import com.example.pam_app.utils.schedulers.SchedulerProvider;

public class ProductionContainer implements Container {

    private final Module module;

    private SchedulerProvider schedulerProvider;
    private BucketRepository bucketRepository;
    private IncomeRepository incomeRepository;
    private LanguagesRepository languagesRepository;
    private WallyDatabase wallyDatabase;
    private BucketDao bucketDao;
    private IncomeDao incomeDao;
    private BucketMapper bucketMapper;
    private IncomeMapper incomeMapper;

    public ProductionContainer(final Context context) {
        this.module = new Module(context);
    }

    @Override
    public Context getApplicationContext() {
        return module.getApplicationContext();
    }

    @Override
    public SchedulerProvider getSchedulerProvider() {
        if (schedulerProvider == null) {
            schedulerProvider = module.provideSchedulerProvider();
        }
        return schedulerProvider;
    }

    @Override
    public BucketRepository getBucketRepository() {
        if (bucketRepository == null) {
            bucketRepository = module.provideBucketRepository(getBucketDao(), getBucketMapper());
        }
        return bucketRepository;
    }

    @Override
    public IncomeRepository getIncomeRepository() {
        if (incomeRepository == null) {
            incomeRepository = module.provideIncomeRepository(getIncomeDao(), getIncomeMapper());
        }
        return incomeRepository;
    }

    @Override
    public LanguagesRepository getLanguageRepository() {
        if (languagesRepository == null) {
            languagesRepository = module.provideLanguageRepository();
        }
        return languagesRepository;
    }

    private WallyDatabase getWallyDatabase() {
        if (wallyDatabase == null) {
            wallyDatabase = module.provideWallyDatabase();
        }
        return wallyDatabase;
    }

    private BucketDao getBucketDao() {
        if (bucketDao == null) {
            bucketDao = module.provideBucketDao(getWallyDatabase());
        }
        return bucketDao;
    }

    private IncomeDao getIncomeDao() {
        if (incomeDao == null) {
            incomeDao = module.provideIncomeDao(getWallyDatabase());
        }
        return incomeDao;
    }

    private BucketMapper getBucketMapper() {
        if (bucketMapper == null) {
            bucketMapper = module.provideBucketMapper();
        }
        return bucketMapper;
    }

    private IncomeMapper getIncomeMapper() {
        if (incomeMapper == null) {
            incomeMapper = module.provideIncomeMapper();
        }
        return incomeMapper;
    }
}

File: app/src/main/java/com/example/pam_app/di/ProductionContainer.kt
package com.example.pam_app.di

import android.content.Context
import com.example.pam_app.db.BucketDao
import com.example.pam_app.db.IncomeDao
import com.example.pam_app.db.WallyDatabase
import com.example.pam_app.repository.*
import com.example.pam_app.utils.schedulers.SchedulerProvider

class ProductionContainer(context: Context?) : Container {
    private val module: Module
    override var schedulerProvider: SchedulerProvider? = null
        get() {
            if (field == null) {
                field = module.provideSchedulerProvider()
            }
            return field
        }
        private set
    override var bucketRepository: BucketRepository? = null
        get() {
            if (field == null) {
                field = module.provideBucketRepository(bucketDao, bucketMapper)
            }
            return field
        }
        private set
    override var incomeRepository: IncomeRepository? = null
        get() {
            if (field == null) {
                field = module.provideIncomeRepository(incomeDao, incomeMapper)
            }
            return field
        }
        private set
    private var languagesRepository: LanguagesRepository? = null
    private var wallyDatabase: WallyDatabase? = null
        get() {
            if (field == null) {
                field = module.provideWallyDatabase()
            }
            return field
        }
    private var bucketDao: BucketDao? = null
        get() {
            if (field == null) {
                field = module.provideBucketDao(wallyDatabase)
            }
            return field
        }
    private var incomeDao: IncomeDao? = null
        get() {
            if (field == null) {
                field = module.provideIncomeDao(wallyDatabase)
            }
            return field
        }
    private var bucketMapper: BucketMapper? = null
        get() {
            if (field == null) {
                field = module.provideBucketMapper()
            }
            return field
        }
    private var incomeMapper: IncomeMapper? = null
        get() {
            if (field == null) {
                field = module.provideIncomeMapper()
            }
            return field
        }
    override val applicationContext: Context
        get() = module.applicationContext

    override val languageRepository: LanguagesRepository?
        get() {
            if (languagesRepository == null) {
                languagesRepository = module.provideLanguageRepository()
            }
            return languagesRepository
        }

    init {
        module = Module(context)
    }
}